src/java/org/apache/cassandra/cql3/Marker.java

```diff
@@ -0,0 +1,147 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.cql3;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import org.apache.cassandra.cql3.functions.Function;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.CollectionType;
+import org.apache.cassandra.db.marshal.ListType;
+import org.apache.cassandra.db.marshal.MapType;
+import org.apache.cassandra.db.marshal.SetType;
+import org.apache.cassandra.db.marshal.TupleType;
+import org.apache.cassandra.db.marshal.UserType;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+import static org.apache.cassandra.cql3.statements.RequestValidations.checkFalse;
+
+/**
+ * A single bind marker.
+ */
+public final class Marker extends Term.NonTerminal
+{
+    private final int bindIndex;
+    private final ColumnSpecification receiver;
+
+    private Marker(int bindIndex, ColumnSpecification receiver)
+    {
+        this.bindIndex = bindIndex;
+        this.receiver = receiver;
+    }
+
+    public void collectMarkerSpecification(VariableSpecifications boundNames)
+    {
+        boundNames.add(bindIndex, receiver);
+    }
+
+    public boolean containsBindMarker()
+    {
+        return true;
+    }
+
+    public void addFunctionsTo(List<Function> functions)
+    {
+    }
+
+    public Term.Terminal bind(QueryOptions options) throws InvalidRequestException
+    {
+        try
+        {
+            ByteBuffer bytes = options.getValues().get(bindIndex);
+            if (bytes == null)
+                return null;
+            if (bytes == ByteBufferUtil.UNSET_BYTE_BUFFER)
+            {
+                checkFalse(receiver.type.isTuple(), "Invalid unset value for tuple %s", receiver.name);
+                return Constants.UNSET_VALUE;
+            }
+
+            return toTerminal(receiver.type, bytes);
+        }
+            catch (MarshalException e)
+        {
+            throw new InvalidRequestException(e.getMessage(), e);
+        }
+    }
+
+    private static Terminal toTerminal(AbstractType<?> type, ByteBuffer bytes)
+    {
+        if (type.isCollection())
+        {
+            switch (((CollectionType) type).kind)
+            {
+                case LIST:
+                    return Lists.Value.fromSerialized(bytes, (ListType<?>) type);
+                case SET:
+                    return Sets.Value.fromSerialized(bytes, (SetType<?>) type);
+                case MAP:
+                    return Maps.Value.fromSerialized(bytes, (MapType<?, ?>) type);
+                default:
+                    throw new AssertionError();
```

Comment:
when will this AssertionError be throw ?

Reply:
I changed the logic.